crypto/src/hash/rescue_prime/mod.rs

```diff
@@ -0,0 +1,816 @@
+use crate::alloc::vec::Vec;
+use core::iter;
+use lambdaworks_math::field::element::FieldElement;
+use lambdaworks_math::field::fields::u64_goldilocks_field::Goldilocks64Field;
+
+mod parameters;
+mod utils;
+
+use parameters::*;
+use utils::*;
+
+type Fp = FieldElement<Goldilocks64Field>;
+
+#[derive(Clone)]
+#[allow(dead_code)]
+enum MdsMethod {
+    MatrixMultiplication,
+    Ntt,
+    Karatsuba,
+}
+#[warn(dead_code)]
+pub struct RescuePrimeOptimized<const SECURITY_LEVEL: usize, const NUM_FULL_ROUNDS: usize> {
+    m: usize,
+    capacity: usize,
+    rate: usize,
+    round_constants: Vec<Fp>,
+    mds_matrix: Vec<Vec<Fp>>,
+    mds_vector: Vec<Fp>,
+    mds_method: MdsMethod,
+}
+
+impl<const SECURITY_LEVEL: usize, const NUM_FULL_ROUNDS: usize> Default
+    for RescuePrimeOptimized<SECURITY_LEVEL, NUM_FULL_ROUNDS>
+{
+    fn default() -> Self {
+        Self::new(MdsMethod::MatrixMultiplication)
+    }
+}
+
+impl<const SECURITY_LEVEL: usize, const NUM_FULL_ROUNDS: usize>
+    RescuePrimeOptimized<SECURITY_LEVEL, NUM_FULL_ROUNDS>
+{
+    fn new(mds_method: MdsMethod) -> Self {
+        assert!(SECURITY_LEVEL == 128 || SECURITY_LEVEL == 160);
+
+        let (m, capacity) = if SECURITY_LEVEL == 128 {
+            (12, 4)
+        } else {
+            (16, 6)
+        };
+        let rate = m - capacity;
+
+        let mds_vector = Self::get_mds_vector(m);
+        let mds_matrix = Self::generate_circulant_matrix(&mds_vector);
+        let round_constants = match SECURITY_LEVEL {
+            128 => get_round_constants(SecurityLevel::Sec128),
+            160 => get_round_constants(SecurityLevel::Sec160),
+            _ => panic!("Unsupported security level"),
+        };
+
+        Self {
+            m,
+            capacity,
+            rate,
+            round_constants: round_constants.to_vec(),
+            mds_matrix,
+            mds_vector,
+            mds_method,
+        }
+    }
+
+    pub fn apply_inverse_sbox(state: &mut [Fp]) {
+        for x in state.iter_mut() {
+            *x = x.pow(ALPHA_INV);
+        }
+    }
+
+    fn get_mds_vector(m: usize) -> Vec<Fp> {
+        match m {
+            12 => vec![7, 23, 8, 26, 13, 10, 9, 7, 6, 22, 21, 8]
+                .into_iter()
+                .map(Fp::from)
+                .collect(),
+            16 => vec![
+                256, 2, 1073741824, 2048, 16777216, 128, 8, 16, 524288, 4194304, 1, 268435456, 1,
+                1024, 2, 8192,
+            ]
+            .into_iter()
+            .map(Fp::from)
+            .collect(),
+            _ => panic!("Unsupported state size"),
+        }
+    }
+
+    fn generate_circulant_matrix(mds_vector: &[Fp]) -> Vec<Vec<Fp>> {
+        let m = mds_vector.len();
```

Comment:
Does this `m` have to do something with the `m` that represents a length that can only be `12` or `16`? If so, don't we need to check its size? And if not, I'd use a more descriptive name for it.

Reply:
Yes, the matrix is a square matrix with `m x m` dimension. However, it's now precomputed, so this not used anymore